sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -378,4 +395,183 @@ object DateTimeUtils {
+    getYearAndDayInYear(date)._2
+  }
+
+  /**
+   * Returns the year value for the given date. The date is expressed in days
+   * since 1.1.1970.
+   */
+  def getYear(date: Int): Int = {
+    getYearAndDayInYear(date)._1
+  }
+
+  /**
+   * Returns the quarter for the given date. The date is expressed in days
+   * since 1.1.1970.
+   */
+  def getQuarter(date: Int): Int = {
+    val (year, dayInYear) = getYearAndDayInYear(date)
+    val leap = if (isLeapYear(year)) 1 else 0
+    if (dayInYear <= 90 + leap) {
+      1
+    } else if (dayInYear <= 181 + leap) {
+      2
+    } else if (dayInYear <= 273 + leap) {
+      3
+    } else {
+      4
+    }
+  }
+
+  /**
+   * Returns the month value for the given date. The date is expressed in days
+   * since 1.1.1970. January is month 1.
+   */
+  def getMonth(date: Int): Int = {
+    val (year, dayInYear) = getYearAndDayInYear(date)
+    val leap = if (isLeapYear(year)) 1 else 0
+    if (dayInYear <= 31) {
+      1
+    } else if (dayInYear <= 59 + leap) {
+      2
+    } else if (dayInYear <= 90 + leap) {
+      3
+    } else if (dayInYear <= 120 + leap) {
+      4
+    } else if (dayInYear <= 151 + leap) {
+      5
+    } else if (dayInYear <= 181 + leap) {
+      6
+    } else if (dayInYear <= 212 + leap) {
+      7
+    } else if (dayInYear <= 243 + leap) {
+      8
+    } else if (dayInYear <= 273 + leap) {
+      9
+    } else if (dayInYear <= 304 + leap) {
+      10
+    } else if (dayInYear <= 334 + leap) {
+      11
+    } else {
+      12
+    }
+  }
+
+  /**
+   * Returns the 'day of month' value for the given date. The date is expressed in days
+   * since 1.1.1970.
+   */
+  def getDayOfMonth(date: Int): Int = {
+    val (year, dayInYear) = getYearAndDayInYear(date)
+    val leap = if (isLeapYear(year)) 1 else 0
+    if (dayInYear <= 31) {
+      dayInYear
+    } else if (dayInYear <= 59 + leap) {
+      dayInYear - 31
+    } else if (dayInYear <= 90 + leap) {
+      dayInYear - 59 - leap
+    } else if (dayInYear <= 120 + leap) {
+      dayInYear - 90 - leap
+    } else if (dayInYear <= 151 + leap) {
+      dayInYear - 120 - leap
+    } else if (dayInYear <= 181 + leap) {
+      dayInYear - 151 - leap
+    } else if (dayInYear <= 212 + leap) {
+      dayInYear - 181 - leap
+    } else if (dayInYear <= 243 + leap) {
+      dayInYear - 212 - leap
+    } else if (dayInYear <= 273 + leap) {
+      dayInYear - 243 - leap
+    } else if (dayInYear <= 304 + leap) {
+      dayInYear - 273 - leap
+    } else if (dayInYear <= 334 + leap) {
+      dayInYear - 304 - leap
+    } else {
+      dayInYear - 334 - leap
+    }
+  }
+
+  /**
+   * Returns the week number for the given date. The date is expressed in days since
+   * 1.1.1970. The first week of the has to hav eat least 4 days. The first day of a
+   * week is Monday.
+   */
+  def getWeekOfYear(date: Int): Int = {
+    calendar.setTimeInMillis(date * MILLIS_PER_DAY)
+    calendar.get(Calendar.WEEK_OF_YEAR)
```

Comment:
sorry for calendar, we should put it back into the expression itself. we can implement a more efficient version later.